### PR TITLE
Multi context refactor

### PR DESCRIPTION
**Pull request by Revolyssup**
**Description**

This PR is Work in progress

**Notes for Reviewers**

The tests will fail because this PR changes few gRPC calls which requires a new release of adapter-library with changes.
[adapter-library PR](https://github.com/meshery/meshery-adapter-library/pull/85) 
**[Signed commits](../CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

**Comment by Revolyssup** [comment]
@gr455  This kind of refactoring will be needed across all adapters. Hope you will champion this. There still might be few changes here, but the basic idea is same that instead of using(or having) a pre-instantiated context, use the one from gRPC call like I explained you that day. I will signal you when this is up for review.

**Comment by gr455** [comment]
> @gr455 This kind of refactoring will be needed across all adapters. Hope you will champion this. There still might be few changes here, but the basic idea is same that instead of using(or having) a pre-instantiated context, use the one from gRPC call like I explained you that day. I will signal you when this is up for review.

Sounds good! Let me give this a look

File: istio/addons.go [hunk 1]
```diff
@@ -27,42 +28,56 @@ func (istio *Istio) installAddon(namespace string, del bool, service string, pat
...
+	var wg sync.WaitGroup
+	var errs []error
+	for _, k8sconfig := range kubeconfigs {
+		wg.Add(1)
+		go func(k8sconfig string) {
+			defer wg.Done()
+			mclient, err := mesherykube.New([]byte(k8sconfig))
 			if err != nil {
-				return st, ErrAddonFromTemplate(err)
+				errs = append(errs, err)
```
Comment (gr455) [inline istio/addons.go:39]: Might need a mutex for this operation?

**Comment by gr455** [comment]
Trying to review this but getting
```
go mod tidy; \
DEBUG=true GOPROXY=direct GOSUMDB=off go run main.go
# github.com/layer5io/meshery-istio/build
build/config.go:29:3: unknown field 'CrdFilter' in struct literal of type manifests.Config
build/config.go:35:3: unknown field 'ExtractCrds' in struct literal of type manifests.Config
make: *** [Makefile:22: run] Error 2
Getting this on your istio pr. I rebased it with master, still shows up
```

Errors show up even after rebasing with master

File: istio/install.go [hunk 2]
```diff
@@ -84,69 +85,100 @@ func (istio *Istio) installIstio(del, useBin bool, version, namespace string, pr
...
+				return
+			}
+			err = kClient.ApplyHelmChart(mesherykube.ApplyHelmChartConfig{
+				LocalPath:       path.Join(downloadLocation, dirName, "manifests/charts/base"),
+				Namespace:       "istio-system",
+				Action:          act,
+				CreateNamespace: true,
+			})
+			if err != nil {
+				errs = append(errs, err)
```
[the remark this answers is not in the document]
Comment (Revolyssup) [reply istio/install.go:118]: I dont think its necessary as I wrap the errors at the end when I merge all errors.

[the remark this answers is not in the document]
Comment (gr455) [reply istio/install.go:118]: Gotcha :+1:

File: istio/install.go [hunk 3]
```diff
@@ -84,69 +85,100 @@ func (istio *Istio) installIstio(del, useBin bool, version, namespace string, pr
...
-		LocalPath:       path.Join(downloadLocation, dirName, "manifests/charts/gateways/istio-egress"),
-		Namespace:       "istio-system",
-		Action:          act,
-		CreateNamespace: true,
-	})
-	if err != nil {
-		return ErrApplyHelmChart(err)
+	return ErrApplyHelmChart(mergeErrors(errs))
+}
+func mergeErrs(errs ...error) (error error) {
```
Comment (gr455) [inline istio/install.go:173]:
```suggestion
	return ErrApplyHelmChart(mergeErrors(errs))
}
func mergeErrors(errs ...error) (error error) {
```
Typo?

File: istio/sample_apps.go [hunk 4]
```diff
@@ -39,27 +41,43 @@ func (istio *Istio) patchWithEnvoyFilter(namespace string, del bool, app string,
 	if err != nil {
 		return st, ErrEnvoyFilter(err)
 	}
+	var wg sync.WaitGroup
+	var errs []error
+	for _, k8sconfig := range kubeconfigs {
+		wg.Add(1)
```
Comment (gr455) [inline istio/sample_apps.go:47]: No `wg.Done()` for this

Comment (Revolyssup) [reply istio/sample_apps.go:47]: Geez, this same code was being copy pasted so many times that I am doing these blunders. Thanks for pointing it out. Please ping if you see any of these

File: istio/install.go (hunk 3 above)
Comment (leecalcote) [reply istio/install.go]: I think that this has been resolved?

**Comment by leecalcote** [comment]
Rerunning build checks - https://github.com/meshery/meshery-istio/actions/runs/2477284180

**Pull request merged**